### PR TITLE
Attempt to query device count via NVML

### PR DESCRIPTION
**Pull request by awaelchli**
## What does this PR do?

Redo of #14319 
Note: This is an enhancement/improvement of the preivous bugfix, not addressing any new bugs. 
- User does not need to set an environment variable
- Faster: no new processes need to be launched in order to evaluate the functions

This adopts the solution in https://github.com/pytorch/pytorch/pull/84879 for Lightning when running with PyTorch < 1.13. 
A minor follow-up is still pending: https://github.com/pytorch/pytorch/pull/85024

Validated with nightly pytorch 1.13 locally, using this simple script:

```py
import torch
import torch.multiprocessing as mp
from lightning_lite.utilities.imports import _TORCH_GREATER_EQUAL_1_13
from lightning_lite.utilities.device_parser import num_cuda_devices, is_cuda_available


def worker(rank):
    print("successfully forked", rank)
    torch.cuda.set_device(rank)


def run():
    print("torch version", torch.__version__)
    print("greater than 1.13?", _TORCH_GREATER_EQUAL_1_13)

    # old function
    torch.cuda.device_count()
    # torch.cuda.is_available()

    # new function
    # print("num_cuda_devices:", num_cuda_devices())
    # print("available:", is_cuda_available())

    mp.start_processes(worker, nprocs=2, start_method="fork")


if __name__ == "__main__":
    run()

```

## Before submitting
- [x] Was this discussed/approved via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [x] Did you make sure to update the documentation with your changes? (if necessary)
- [ ] Did you write any new necessary tests? (not for typos and docs)
- [ ] Did you verify new and existing tests pass locally with your changes?
- [ ] Did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

 - [ ] Is this pull request ready for review? (if not, please submit in draft mode)
 - [x] Check that all items from **Before submitting** are resolved
 - [x] Make sure the title is self-explanatory and the description concisely explains the PR
 - [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?
I made sure I had fun coding 🙃

cc @borda @akihironitta

### CODE REVIEW

File: src/lightning_lite/utilities/device_parser.py [hunk 1]
```diff
@@ -286,16 +287,30 @@ def _parse_tpu_cores_str(tpu_cores: str) -> Union[int, List[int]]:
...
+    try:
+        nvml_h = CDLL("libnvidia-ml.so.1")
+    except OSError:
+        # Can't load NVML shared libraries, fall back to torch.cuda
         return torch.cuda.device_count()
-    with multiprocessing.get_context("fork").Pool(1) as pool:
-        return pool.apply(torch.cuda.device_count)
+
+    rc = nvml_h.nvmlInit()
+    assert rc == 0
```
Comment (carmocca) [inline src/lightning_lite/utilities/device_parser.py:307]: I would switch these assertions for gotos into exception handling code that falls back to something sensible, `device_count` or the previous `get_context` solution

**Review by carmocca — APPROVED** [review]
Great that this was fixed!

**Review by otaj — APPROVED** [review]

**Review by rohitgr7 — APPROVED** [review]

**Review by justusschock — APPROVED** [review]
